src/textprops.rs

```diff
@@ -0,0 +1,110 @@
+use crate::{
+    core::{
+        cons::Cons,
+        gc::Context,
+        object::{Gc, ListType, Object},
+    },
+    fns::eq,
+};
+use anyhow::Result;
+
+#[allow(dead_code)]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum PropertySetType {
+    Replace,
+    Prepend,
+    Append,
+}
+
+// Add the properties of PLIST to the interval I, or set
+//    the value of I's property to the value of the property on PLIST
+//    if they are different.
+
+//    OBJECT should be the string or buffer the interval is in.
+
+//    If DESTRUCTIVE, the function is allowed to reuse list values in the
+//    properties.
+
+//    Return true if this changes I (i.e., if any members of PLIST
+//    are actually added to I's plist)
+pub fn add_properties<'ob>(
+    plist: Object<'ob>,
+    plist_i: Object<'ob>,
+    _set_type: PropertySetType,
+    _destructive: bool,
+    cx: &'ob Context,
+) -> Result<bool> {
+    // TODO return type
+    let mut changed = false;
+    let Ok(plist) = Gc::<ListType>::try_from(plist) else { return Ok(false) };
+    let Ok(mut plist_i) = Gc::<ListType>::try_from(plist_i) else { return Ok(false) };
+    let mut iter = plist.elements();
+    // iterate through plist, finding key1 and val1
+    while let Some(key1) = iter.next() {
+        let key1 = key1?;
+        let Some(val1) = iter.next() else { return Ok(changed) };
+        let mut found = false;
+
+        let mut iter_i = plist_i.conses();
+        // iterate through i's plist, finding (key2, val2) and set val2 if key2 == key1;
+        while let Some(key2_cons) = iter_i.next() {
+            let Some(val2_cons) = iter_i.next() else { return Ok(changed) };
+            if eq(key1, key2_cons?.car()) {
+                // TODO this should depend on set_type
+                val2_cons?.set_car(val1?)?;
+                changed = true;
+                found = true;
+                break;
+            }
+        }
+        // if no key1 found, append them
+        if !found {
+            let pl = plist_i.untag();
+            changed = true;
+            let new_cons: Gc<ListType<'ob>> = Cons::new(key1, Cons::new(val1?, pl, cx), cx).into();
+            plist_i = new_cons; // TODO this does not work
```

Comment:
How should I turn a Nil Cons to a contented one?

Reply:
anything that is a subset of an `Object` can be converted to an `Object` via `into`. So if you wanted to update `plist_i` you can first make it mutable

```rust
pub fn add_properties<'ob>(
...
mut plist_i: Object<'ob>,
```
then update it:

```rust
let new_cons: List = Cons::new(key1, Cons::new(val1?, pl, cx), cx).into();
 plist_i = new_cons.into();
```

or even just like this

```rust
plist_i = Cons::new(key1, Cons::new(val1?, pl, cx), cx).into();
```

Reply:
thanks for this, actually it is just how I did. now i'm passing references `plist: &mut Object<'ob>` to modify the plist inplace.

Reply:
Note that `Object` is just a tagged pointer, so you can't modify the underlying data via `&mut` (it would be similar to having `&mut &String`, where you have a *mutable* reference to an *immutable* string). if `plist_i` is a `Cons`, than you can use `.set_car()` or `.set_cdr()`. 

Though looking at what you are trying to do, since we want to preprend key1. But that will change the head of the list, so we can't do it mutably. Instead we should return a new `plist_i` and then reassign it outside of `add_properties`. Either that or *append* key1 (instead of preprend) and that will modify the existing list without changing the head.